File: java/jdk_practice/src/main/java/design_pattern/FactoryMethod.java
package design_pattern;

/**
 * @author xiaoqiang
 * @date 2020/8/27-10:48
 */
interface Operator {
    void SetA(int a);

    void SetB(int b);

    int Result();
}

interface OperactorFactory {
    Operator Create();
}

//基类抽象 实现 SetA，SetB的公共方法
abstract class BaseOperator implements Operator {
    int a;
    int b;

    @Override
    public void SetA(int a) {
        this.a = a;
    }

    @Override
    public void SetB(int b) {
        this.b = b;
    }
}

//加法
class PlusOperator extends BaseOperator {
    @Override
    public int Result() {
        return a + b;
    }
}

//工厂
class PlusOperatorFactory implements OperactorFactory {
    @Override
    public Operator Create() {
        return new PlusOperator();
    }
}

public class FactoryMethod {
    public static void main(String[] args) {
        //可以通过其他的加法之类的 根据策略模式返回对应的工厂
        OperactorFactory factory = new PlusOperatorFactory();
        Operator operator = factory.Create();
        operator.SetA(1);
        operator.SetB(2);
        System.out.println(operator.Result());
    }
}

File: go/design_pattern/factory_method/factorymethod.go
package factory_method

//工厂方法模式使用子类的方式延迟生成对象到子类中实现。
//Go中不存在继承 所以使用匿名组合来实现

//Operator 是被封装的实际类接口
type Operator interface {
	SetA(int)
	SetB(int)
	Result() int
}

//OperatorFactory 是工厂接口
type OperatorFactory interface {
	//创建一个
	Create() Operator
}

//OperatorBase 是Operator 接口实现的基类，封装公用方法 Set
type OperatorBase struct {
	a, b int
}

func (o *OperatorBase) SetA(a int) {
	o.a = a
}
func (o *OperatorBase) SetB(b int) {
	o.b = b
}

//加法, 匿名字段继承
type PlusOperator struct {
	*OperatorBase
}

func (p *PlusOperator) Result() int {
	return p.a + p.b
}

//PlusOperatorFactory 是 PlusOperator 的工厂类
type PlusOperatorFactory struct{}

//创建自己的一个复制
func (PlusOperatorFactory) Create() Operator {
	return &PlusOperator{
		OperatorBase: &OperatorBase{},
	}
}

type MinusOperator struct {
	*OperatorBase
}

//创建

//获取结果
func (m *MinusOperator) Result() int {
	//获取结果减法
	return m.a - m.b
}

type MinusOperatorFactory struct {
}

func (m MinusOperatorFactory) Create() Operator {
	return &MinusOperator{}
}
